### PR TITLE
Update error codes of Patch operation-IDP Management REST API

### PR DESCRIPTION
**Pull request by ChanikaRuchini**
## Purpose
fixes - [#12499](https://github.com/wso2/product-is/issues/12499)

**Updated the following error status codes.**

1. JWKS URI update

- PATCH REMOVE non existing property - 404 not found
- PATCH REPLACE non existing property - 404 not found

2. Certificate update

- PATCH ADD existing certificate  - 409 conflict
- PATCH ADD with invalid path index - 400 bad requet
- PATCH REMOVE non existing certifiacte - 404 not found
- PATCH REPLACE non existing certifiacte - 404 not found
- PATCH REPLACE certifiacte with other exiting certifiacte - 409 conflict

#### Merge after [identity-api-server#308](https://github.com/wso2/identity-api-server/pull/308)

### CODE REVIEW

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java [hunk 1]
```diff
@@ -2733,55 +2742,124 @@ private void processPatchRequest(List<Patch> patchRequest, IdentityProvider idpT
...
-                    throw handleException(Response.Status.BAD_REQUEST, Constants.ErrorMessage
-                            .ERROR_CODE_INVALID_INPUT, "Invalid index in 'path' attribute");
-                }
-                if (ArrayUtils.isNotEmpty(idpToUpdate.getCertificateInfoArray())) {
-                    if (index > idpToUpdate.getCertificateInfoArray().length) {
-                        throw handleException(Response.Status.BAD_REQUEST, Constants.ErrorMessage
-                                .ERROR_CODE_INVALID_INPUT, "Invalid index in 'path' attribute");
+            } else if (operation == Patch.OperationEnum.ADD) {
+                if (isCertificateUpdateRequest) {
+
```
Comment (ChamathNS) [inline components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:2755]: redundant blank line

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java [hunk 2]
```diff
@@ -2733,55 +2742,124 @@ private void processPatchRequest(List<Patch> patchRequest, IdentityProvider idpT
...
+                    List<String> certificates = new ArrayList<>();
+                    int index = Integer.parseInt(path.split(Constants.PATH_SEPERATOR)[3]);
```
Comment (ChamathNS) [inline components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:2757]: Wouldn't this throw a NumberFormatException?

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java [hunk 3]
```diff
@@ -2733,55 +2742,124 @@ private void processPatchRequest(List<Patch> patchRequest, IdentityProvider idpT
...
+                    if (index != idpToUpdate.getCertificateInfoArray().length) {
+                        throw handleException(Response.Status.BAD_REQUEST,
+                                Constants.ErrorMessage.ERROR_CODE_INVALID_INPUT, "Invalid index in 'path' attribute");
+                    }
+                    if (ArrayUtils.isNotEmpty(idpToUpdate.getCertificateInfoArray())) {
+                        for (CertificateInfo certInfo : idpToUpdate.getCertificateInfoArray()) {
+                            certificates.add(base64Decode(certInfo.getCertValue()));
+                        }
```
Comment (ChamathNS) [inline components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:2765]: Let's try to use Java 8 forEach :)

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java [hunk 4]
```diff
@@ -2733,55 +2742,124 @@ private void processPatchRequest(List<Patch> patchRequest, IdentityProvider idpT
...
+                    List<IdentityProviderProperty> idpNewProperties = new ArrayList<>();
+                    for (IdentityProviderProperty propertyDTO : propertyDTOS) {
+                        // Add properties to new list omitting the JWKS URI property.
+                        if (!Constants.JWKS_URI.equals(propertyDTO.getName())) {
+                            idpNewProperties.add(propertyDTO);
+                        }
+                    }
+                    idpToUpdate.setIdpProperties(idpNewProperties.toArray(new IdentityProviderProperty[0]));
+                } else if (Constants.CERTIFICATE_JWKSURI_PATH.equals(path)) {
+
```
Comment (ChamathNS) [inline components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:2795]: Redundant blank line

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java [hunk 5]
```diff
@@ -2733,55 +2742,124 @@ private void processPatchRequest(List<Patch> patchRequest, IdentityProvider idpT
...
+                                "Cannot remove JWKS URI as it does not exist");
                     }
-                    certificates.remove(index);
+
+                    idpToUpdate.setIdpProperties(idpNewProperties.toArray(new IdentityProviderProperty[0]));
                 } else {
-                    throw handleException(Response.Status.BAD_REQUEST, Constants.ErrorMessage
-                            .ERROR_CODE_INVALID_INPUT, "Invalid index in 'path' attribute");
+                    throw handleException(Response.Status.BAD_REQUEST,
+                            Constants.ErrorMessage.ERROR_CODE_INVALID_INPUT, null);
```
Comment (ChamathNS) [inline components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:2862]: This method is too long. Better if we can refactor it by extracting out private methods

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java (hunk 1 above)
Comment (ChanikaRuchini) [reply components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:2755]: Related to the dependent PR

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java (hunk 2 above)
Comment (ChanikaRuchini) [reply components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:2757]: Related to the dependent PR

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java (hunk 3 above)
Comment (ChanikaRuchini) [reply components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:2765]: Related to the dependent PR

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java (hunk 4 above)
Comment (ChanikaRuchini) [reply components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:2795]: Related to the dependent PR

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java (hunk 5 above)
Comment (ChanikaRuchini) [reply components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:2862]: Related to the dependent PR

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java [hunk 6]
```diff
@@ -2862,8 +2875,14 @@ private void patchIdpProperties(IdentityProvider identityProvider, String proper
...
             if (propertyName.equals(propertyDTO.getName())) {
                 propertyDTO.setValue(propertyValue);
+                return;
             }
         }
+
+        if (Constants.JWKS_URI.equals(propertyName)) {
+            throw handleException(Response.Status.NOT_FOUND, Constants.ErrorMessage.ERROR_CODE_ERROR_UPDATING_IDP,
+                    "Cannot replace JWKS URI as it does not exist");
+        }
```
Comment (ashendes) [inline components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:2885]: This should be handled in a generic way for any missing property.

Comment (ChanikaRuchini) [reply components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java]: Fixed

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java [hunk 7]
```diff
@@ -2698,8 +2698,17 @@ private void processPatchRequest(List<Patch> patchRequest, IdentityProvider idpT
                                         Constants.ErrorMessage.ERROR_CODE_INVALID_CERTIFICATE_FORMAT, null);
                             }
                         }
+                        if (certificates.contains(value)) {
+                            throw handleException(Response.Status.CONFLICT,
+                                    Constants.ErrorMessage.ERROR_CODE_ERROR_UPDATING_IDP,
+                                    "Cannot replace certificate as this certificate is already exists.");
```
Comment (ChamathNS) [inline components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:2704]:
```suggestion
                                    "Cannot replace certificate as this certificate already exists.");
```

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java [hunk 8]
```diff
@@ -2698,8 +2698,17 @@ private void processPatchRequest(List<Patch> patchRequest, IdentityProvider idpT
...
+                        }
                         certificates.set(index, value);
                         idpToUpdate.setCertificate(base64Encode(StringUtils.join(certificates, "")));
+                    } else if (ArrayUtils.isEmpty(idpToUpdate.getCertificateInfoArray()) ||
+                            index >= idpToUpdate.getCertificateInfoArray().length) {
+                        throw handleException(Response.Status.NOT_FOUND, Constants.ErrorMessage
+                                .ERROR_CODE_ERROR_UPDATING_IDP, "Cannot replace certificate as it does not exist");
```
Comment (ChamathNS) [inline components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:2711]:
```suggestion
                                .ERROR_CODE_ERROR_UPDATING_IDP, "Cannot replace certificate as it does not exist.");
```

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java [hunk 9]
```diff
@@ -2768,6 +2772,11 @@ private void processPatchRequest(List<Patch> patchRequest, IdentityProvider idpT
                                     Constants.ErrorMessage.ERROR_CODE_INVALID_CERTIFICATE_FORMAT, null);
                         }
                     }
+                    if (certificates.contains(value)) {
+                        throw handleException(Response.Status.CONFLICT,
+                                Constants.ErrorMessage.ERROR_CODE_ERROR_UPDATING_IDP,
+                                "Cannot add certificate as it already exists");
```
Comment (ChamathNS) [inline components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:2778]: LEt's use sentence case. Fix other instances as well

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java (hunk 7 above)
Comment (ChanikaRuchini) [reply components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java]: Resolved

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java (hunk 8 above)
Comment (ChanikaRuchini) [reply components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java]: Resolved

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java (hunk 9 above)
Comment (ChanikaRuchini) [reply components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java]: Resolved

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java [hunk 10]
```diff
@@ -2768,6 +2772,11 @@ private void processPatchRequest(List<Patch> patchRequest, IdentityProvider idpT
                                     Constants.ErrorMessage.ERROR_CODE_INVALID_CERTIFICATE_FORMAT, null);
                         }
                     }
+                    if (certificates.contains(value)) {
+                        throw handleException(Response.Status.CONFLICT,
+                                Constants.ErrorMessage.ERROR_CODE_ERROR_UPDATING_IDP,
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:2777]: Can you check the formatting?

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java [hunk 11]
```diff
@@ -2862,8 +2875,12 @@ private void patchIdpProperties(IdentityProvider identityProvider, String proper
         for (IdentityProviderProperty propertyDTO : propertyDTOS) {
             if (propertyName.equals(propertyDTO.getName())) {
                 propertyDTO.setValue(propertyValue);
+                return;
             }
         }
+
+        throw handleException(Response.Status.NOT_FOUND, Constants.ErrorMessage.ERROR_CODE_ERROR_UPDATING_IDP,
+                String.format("Cannot replace  %s as it does not exist", propertyName));
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:2883]:
```suggestion
                String.format("Cannot replace %s as it does not exist", propertyName));
```

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java [hunk 12]
```diff
@@ -2698,8 +2698,17 @@ private void processPatchRequest(List<Patch> patchRequest, IdentityProvider idpT
...
                             }
                         }
+                        if (certificates.contains(value)) {
+                            throw handleException(Response.Status.CONFLICT,
+                                    Constants.ErrorMessage.ERROR_CODE_ERROR_UPDATING_IDP,
+                                    "Cannot replace certificate as this certificate already exists.");
+                        }
                         certificates.set(index, value);
                         idpToUpdate.setCertificate(base64Encode(StringUtils.join(certificates, "")));
+                    } else if (ArrayUtils.isEmpty(idpToUpdate.getCertificateInfoArray()) ||
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:2708]: Can `idpToUpdate` be null?

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java [hunk 13]
```diff
@@ -2698,8 +2698,17 @@ private void processPatchRequest(List<Patch> patchRequest, IdentityProvider idpT
                                         Constants.ErrorMessage.ERROR_CODE_INVALID_CERTIFICATE_FORMAT, null);
                             }
                         }
+                        if (certificates.contains(value)) {
+                            throw handleException(Response.Status.CONFLICT,
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:2702]: Check formatting.

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java (hunk 12 above)
Comment (ChanikaRuchini) [reply components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:2708]: ArrayUtils.isEmpty checks whether the given array is empty or null.

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java (hunk 11 above)
Comment (ChanikaRuchini) [reply components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java]: Resolved

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java (hunk 13 above)
Comment (ChanikaRuchini) [reply components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:2702]: Rechecked with codestyles. no issues found.

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java (hunk 10 above)
Comment (ChanikaRuchini) [reply components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:2777]: Rechecked with codestyles. no issues found.

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java (hunk 12 above)
Comment (ashensw) [reply components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:2708]: It's not about the CertificateInfoArray. If the `idpToUpdate` object is null, won't this cause an NPE?

Comment (ChanikaRuchini) [reply components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:2708]: It is already handled in https://github.com/wso2/identity-api-server/blob/6b17264995093bc2d93df3fe33ed8c901d688c9e/components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java#L230

**Review by ashensw — APPROVED** [review]

**Pull request merged**